File: go/contest_test.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

package bnf

// Whether two dispatch heads CONTEST (can the lexer hand the same input
// to both) decides how deep the dispatcher looks. An answer of "no" that
// is wrong emits one-token entries for a decision that needs two, and the
// first branch commits on the first token and rejects input the other
// branch accepts. Mirrors ts/test/contest.test.js.

import (
	"reflect"
	"strings"
	"testing"

	tabnas "github.com/tabnas/parser/go"
)

func ctLit(s string) *Element {
	return &Element{Kind: KindTerm, Literal: s, CaseSensitive: true, HasCaseSens: true}
}
func ctILit(s string) *Element { return &Element{Kind: KindTerm, Literal: s} }
func ctRef(n string) *Element  { return &Element{Kind: KindRef, Name: n} }

// doc = x ; x = <a> t u / <b> u t ; t = "." "." ; u = "," ","
func ctGrammar(a, b *Element) *Grammar {
	return &Grammar{Productions: []*Production{
		{Name: "doc", Alts: []Sequence{{ctRef("x")}}},
		{Name: "x", Alts: []Sequence{{a, ctRef("t"), ctRef("u")}, {b, ctRef("u"), ctRef("t")}}},
		{Name: "t", Alts: []Sequence{{ctLit("."), ctLit(".")}}},
		{Name: "u", Alts: []Sequence{{ctLit(","), ctLit(",")}}},
	}}
}

// ctDepths is the number of lookahead tokens each open alternate of x
// names. Open is `any`; its elements carry the alternate's `s` either as
// a map entry or as a struct field, depending on the emitter's form.
func ctDepths(t *testing.T, spec *tabnas.GrammarSpec) []int {
	t.Helper()
	rule, ok := spec.Rule["x"]
	if !ok || rule == nil {
		t.Fatal("no rule x")
	}
	v := reflect.ValueOf(rule.Open)
	if !v.IsValid() || v.Kind() != reflect.Slice {
		t.Fatalf("open is %T", rule.Open)
	}
	var out []int
	for i := 0; i < v.Len(); i++ {
		el := v.Index(i)
		for el.Kind() == reflect.Interface || el.Kind() == reflect.Ptr {
			el = el.Elem()
		}
		var s string
		switch el.Kind() {
		case reflect.Map:
			s, _ = el.MapIndex(reflect.ValueOf("s")).Interface().(string)
		case reflect.Struct:
			f := el.FieldByName("S")
			if f.IsValid() {
				s, _ = f.Interface().(string)
			}
		default:
			t.Fatalf("alternate %d is %s", i, el.Kind())
		}
		out = append(out, len(strings.Fields(s)))
	}
	return out
}

func ctParses(t *testing.T, spec *tabnas.GrammarSpec, src string, relex bool) bool {
	t.Helper()
	j := tabnas.Make(tabnas.Options{Lex: &tabnas.LexOptions{Relex: tabnas.Bool(relex)}})
	if err := j.Grammar(spec); err != nil {
		t.Fatal(err)
	}
	_, err := j.Parse(src)
	return err == nil
}

func TestContestWordKeywordAgainstPunctuationContinuation(t *testing.T) {
	spec, err := EmitGrammarSpec(ctGrammar(ctLit("a"), ctLit("a-b")), &ConvertOptions{Tag: "ct", Start: "doc", WordKeywords: true})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
		t.Fatalf("depths %v, want [2 2]", d)
	}
	for _, src := range []string{"a-b,,..", "a..,,"} {
		if !ctParses(t, spec, src, true) {
			t.Errorf("%q should parse", src)
		}
	}
	word, err := EmitGrammarSpec(ctGrammar(ctLit("a"), ctLit("ab")), &ConvertOptions{Tag: "ct", Start: "doc", WordKeywords: true})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, word); d[0] != 1 || d[1] != 1 {
		t.Fatalf("depths %v, want [1 1]: a word continuation keeps the guard", d)
	}
}

func TestContestCaseInsensitiveLiteralsFoldAsTheirMatchersDo(t *testing.T) {
	// A literal with no ASCII letter compiles to an exact fixed token, so
	// Σ and ς never meet; with ASCII letters the matcher folds case.
	spec, err := EmitGrammarSpec(ctGrammar(ctILit("Σ"), ctILit("ς")), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, spec); d[0] != 1 || d[1] != 1 {
		t.Fatalf("depths %v, want [1 1]", d)
	}
	for _, src := range []string{"ς,,..", "Σ..,,"} {
		if !ctParses(t, spec, src, false) {
			t.Errorf("%q should parse", src)
		}
	}
	ascii, err := EmitGrammarSpec(ctGrammar(ctILit("a"), ctILit("A-b")), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, ascii); d[0] != 2 || d[1] != 2 {
		t.Fatalf("depths %v, want [2 2]", d)
	}
	for _, src := range []string{"A-B,,..", "A..,,"} {
		if !ctParses(t, ascii, src, true) {
			t.Errorf("%q should parse", src)
		}
	}
}

func TestContestRegexHeadThatIsNotOneAtom(t *testing.T) {
	spec, err := EmitGrammarSpec(ctGrammar(&Element{Kind: KindRegex, Pattern: "a|b"}, ctLit("b")), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
		t.Fatalf("depths %v, want [2 2]", d)
	}
	for _, src := range []string{"b,,..", "a..,,"} {
		if !ctParses(t, spec, src, true) {
			t.Errorf("%q should parse", src)
		}
	}
}

func TestContestAnyTokenMeetsEveryHead(t *testing.T) {
	spec, err := EmitGrammarSpec(ctGrammar(&Element{Kind: KindToken, Name: "#AA"}, ctLit("b")), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
		t.Fatalf("depths %v, want [2 2]", d)
	}
	for _, src := range []string{"b,,..", "z..,,"} {
		if !ctParses(t, spec, src, false) {
			t.Errorf("%q should parse", src)
		}
	}
}

func TestContestRegexHeadWhoseCoverageIsUnnamed(t *testing.T) {
	// \n, \t and \v are one atom each, but patternCharRanges declines to
	// name what a control escape covers, so nothing says the literal
	// character is not what the pattern matches.
	for _, c := range []struct{ pattern, literal string }{
		{`\n`, "\n"}, {`\t`, "\t"}, {`\v`, "\v"},
	} {
		spec, err := EmitGrammarSpec(ctGrammar(&Element{Kind: KindRegex, Pattern: c.pattern}, ctLit(c.literal)), &ConvertOptions{Tag: "ct", Start: "doc"})
		if err != nil {
			t.Fatalf("%s: %v", c.pattern, err)
		}
		if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
			t.Fatalf("%s: depths %v, want [2 2]", c.pattern, d)
		}
	}
	spec, err := EmitGrammarSpec(ctGrammar(&Element{Kind: KindRegex, Pattern: `\v`}, ctLit("\v")), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	for _, src := range []string{"\v..,,", "\v,,.."} {
		if !ctParses(t, spec, src, true) {
			t.Errorf("%q should parse", src)
		}
	}
}

func TestContestCaseInsensitiveHeadBeyondASCII(t *testing.T) {
	// (?i)[Σ] takes ς, and the coverage folds ASCII letters alone.
	spec, err := EmitGrammarSpec(ctGrammar(&Element{Kind: KindRegex, Pattern: "[Σ]", Flags: "i"}, ctLit("ς")), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
		t.Fatalf("depths %v, want [2 2]", d)
	}
	for _, src := range []string{"ς..,,", "Σ..,,", "ς,,.."} {
		if !ctParses(t, spec, src, true) {
			t.Errorf("%q should parse", src)
		}
	}
	// Within ASCII the folded coverage stays exact: (?i)[b] meets no a.
	ascii, err := EmitGrammarSpec(ctGrammar(&Element{Kind: KindRegex, Pattern: "[b]", Flags: "i"}, ctLit("a")), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, ascii); d[0] != 1 || d[1] != 1 {
		t.Fatalf("depths %v, want [1 1]", d)
	}
}

// ctSemi is ctGrammar with tails no engine matcher can take (a number
// would swallow the `.` of `1..`): t = ";" ";" ; u = "!" "!".
func ctSemi(a, b *Element) *Grammar {
	return &Grammar{Productions: []*Production{
		{Name: "doc", Alts: []Sequence{{ctRef("x")}}},
		{Name: "x", Alts: []Sequence{{a, ctRef("t"), ctRef("u")}, {b, ctRef("u"), ctRef("t")}}},
		{Name: "t", Alts: []Sequence{{ctLit(";"), ctLit(";")}}},
		{Name: "u", Alts: []Sequence{{ctLit("!"), ctLit("!")}}},
	}}
}

func TestContestEngineTokenMeetsWhatItsMatcherCanTake(t *testing.T) {
	// Negotiated lexing runs only the matchers that can produce the token
	// an alternative wants: the number matcher takes a leading digit, the
	// string matcher a quote, and the text matcher any text no fixed
	// literal claims. The four-token dispatch this replaced kept each of
	// these pairs apart; a one-token dispatch did not.
	tk := func(n string) *Element { return &Element{Kind: KindToken, Name: n} }
	rx := func(p string) *Element { return &Element{Kind: KindRegex, Pattern: p} }
	for _, c := range []struct {
		a, b *Element
		text string
	}{
		{tk("#NR"), ctLit("1"), "1"},
		{tk("#ST"), ctLit("'a'"), "'a'"},
		{tk("#TX"), ctILit("let"), "let"},
		{tk("#NR"), tk("#TX"), "1"},
		{tk("#NR"), rx("[0-9]"), "1"},
		{tk("#TX"), rx("[a-z]+"), "let"},
	} {
		spec, err := EmitGrammarSpec(ctSemi(c.a, c.b), &ConvertOptions{Tag: "ct", Start: "doc"})
		if err != nil {
			t.Fatal(err)
		}
		if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
			t.Fatalf("%s: depths %v, want [2 2]", c.text, d)
		}
		for _, src := range []string{c.text + ";;!!", c.text + "!!;;"} {
			if !ctParses(t, spec, src, true) {
				t.Errorf("%q should parse", src)
			}
		}
	}
	// The text matcher defers to a fixed literal, and an emitted grammar
	// lexes no values: those stay one token deep.
	for _, c := range []struct{ a, b *Element }{
		{tk("#TX"), ctLit("let")}, {tk("#VL"), ctLit("true")},
	} {
		spec, err := EmitGrammarSpec(ctSemi(c.a, c.b), &ConvertOptions{Tag: "ct", Start: "doc"})
		if err != nil {
			t.Fatal(err)
		}
		if d := ctDepths(t, spec); d[0] != 1 || d[1] != 1 {
			t.Fatalf("%s: depths %v, want [1 1]", c.a.Name, d)
		}
	}
	// TypeScript and Rust also pin that `\u{1}` is inexact without the u
	// flag. Go's regexp has no `\u{...}` escape, so such a pattern never
	// reaches the predicate here.
}

func TestContestEscapeWhoseCodePointCannotBeReadIsNotExact(t *testing.T) {
	// The head scanner and the coverage reader take an escape the same
	// way: one code point they can name, or unknown. `\u1` and `\x1` are
	// `u` then `1` and `x` then `1` in the JavaScript matcher the
	// canonical runtime emits for, and `\cA`, `\p{L}`, `\k` and a digit
	// escape are a control character, a property, a group or a back
	// reference, none of them the letter after the backslash. Go's regexp
	// compiles only some of these, so they are pinned on the two readers.
	// JavaScript's `\u` is not an RE2 escape at all, so even with its full
	// digits it names nothing here (readEscape reads RE2).
	for _, p := range []string{`\u1`, `\x1`, `\u12`, `\uD83D`, `\u{}`, `\u{4g}`, `\x{110000}`,
		`\cA`, `\p{L}`, `\PL`, `\k<a>`, `\1`, `\0`, `\d`, `\u0041`, `\u{1F600}`} {
		if end := regexHeadAtomEnd(p); end != -1 {
			t.Errorf("regexHeadAtomEnd(%q) = %d, want -1", p, end)
		}
		if r := patternCharRanges(p); r != nil {
			t.Errorf("patternCharRanges(%q) = %v, want unknown", p, r)
		}
	}
	for _, p := range []string{`[\p{L}]`, `[a\1]`, `[\u1]`} {
		if r := patternCharRanges(p); r != nil {
			t.Errorf("patternCharRanges(%q) = %v, want unknown", p, r)
		}
	}
	for _, c := range []struct {
		p   string
		end int
		cp  rune
	}{
		{`\x41`, 4, 'A'}, {`\x{41}`, 6, 'A'}, {`\x{1F600}`, 9, 0x1F600}, {`\.`, 2, '.'},
	} {
		if end := regexHeadAtomEnd(c.p); end != c.end {
			t.Errorf("regexHeadAtomEnd(%q) = %d, want %d", c.p, end, c.end)
		}
		if r := patternCharRanges(c.p); len(r) != 1 || r[0] != (charRange{c.cp, c.cp}) {
			t.Errorf("patternCharRanges(%q) = %v, want %U", c.p, r, c.cp)
		}
	}
	// And through the dispatcher, where the regexp compiles: a property
	// class meets a letter it covers.
	spec, err := EmitGrammarSpec(ctSemi(&Element{Kind: KindRegex, Pattern: `[\p{L}]`}, ctLit("é")),
		&ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
		t.Fatalf(`[\p{L}]: depths %v, want [2 2]`, d)
	}
}

func TestContestEscapeIsReadAsRE2ReadsIt(t *testing.T) {
	// This port compiles every matcher with Go's regexp, so an escape is
	// read as RE2 reads it, not as JavaScript does. `\a` is BEL, which a
	// JavaScript matcher reads as the letter `a`; read as `a`, a head `\a`
	// was held apart from a literal BEL that it takes, and the literal's
	// branch was never reached. The zero-width `\A` and `\z`, the quoting
	// `\Q…\E`, `\C` and letters RE2 does not define name no one code point.
	for _, c := range []struct {
		p   string
		end int
		cp  rune
	}{
		{`\a`, 2, 0x07}, {`\x07`, 4, 0x07}, {`\_`, 2, '_'}, {`\-`, 2, '-'},
	} {
		if end := regexHeadAtomEnd(c.p); end != c.end {
			t.Errorf("regexHeadAtomEnd(%q) = %d, want %d", c.p, end, c.end)
		}
		if r := patternCharRanges(c.p); len(r) != 1 || r[0] != (charRange{c.cp, c.cp}) {
			t.Errorf("patternCharRanges(%q) = %v, want %U", c.p, r, c.cp)
		}
	}
	if r := patternCharRanges(`[\a-\x{0d}]`); len(r) != 1 || r[0] != (charRange{0x07, 0x0D}) {
		t.Errorf(`patternCharRanges([\a-\x{0d}]) = %v, want U+0007-U+000D`, r)
	}
	for _, p := range []string{`\A`, `\z`, `\Qa\E`, `\Q+\E`, `\C`, `\E`, `\e`, `\U00000041`} {
		if end := regexHeadAtomEnd(p); end != -1 {
			t.Errorf("regexHeadAtomEnd(%q) = %d, want -1", p, end)
		}
		if r := patternCharRanges(p); r != nil {
			t.Errorf("patternCharRanges(%q) = %v, want unknown", p, r)
		}
	}
	// Through the dispatcher: each of these heads can take the input the
	// literal takes, so the choice looks two tokens deep and both branches
	// are reached.
	for _, c := range []struct{ pattern, text string }{
		{`\a`, "\a"}, {`[\a]`, "\a"}, {`\a+`, "\a"}, {`\Q+\E`, "+"},
	} {
		spec, err := EmitGrammarSpec(ctSemi(&Element{Kind: KindRegex, Pattern: c.pattern}, ctLit(c.text)),
			&ConvertOptions{Tag: "ct", Start: "doc"})
		if err != nil {
			t.Fatal(err)
		}
		if d := ctDepths(t, spec); d[0] != 2 || d[1] != 2 {
			t.Errorf("%s: depths %v, want [2 2]", c.pattern, d)
		}
		for _, src := range []string{c.text + ";;!!", c.text + "!!;;"} {
			if !ctParses(t, spec, src, true) {
				t.Errorf("%s: %q should parse", c.pattern, src)
			}
		}
	}
	// A range ending on a hex escape that names a surrogate is read as RE2
	// reads it, and laid over the partition beside the class it overlaps.
	// (The TypeScript reader once named no surrogate escape at all, and
	// lost the second branch to the lexer.)
	spec, err := EmitGrammarSpec(ctSemi(&Element{Kind: KindRegex, Pattern: `[\x{0041}-\x{d800}]`},
		&Element{Kind: KindRegex, Pattern: `[\x{0041}-\x{005a}]`}), &ConvertOptions{Tag: "ct", Start: "doc"})
	if err != nil {
		t.Fatal(err)
	}
	for _, src := range []string{"A;;!!", "A!!;;", "b;;!!"} {
		if !ctParses(t, spec, src, false) {
			t.Errorf("%q should parse", src)
		}
	}
}

File: rs/tests/contest_test.rs
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

//! Whether two dispatch heads CONTEST (can the lexer hand the same input
//! to both) decides how deep the dispatcher looks. An answer of "no" that
//! is wrong emits one-token entries for a decision that needs two, and
//! the first branch commits on the first token and rejects input the
//! other branch accepts. Mirrors `ts/test/contest.test.js`.

mod common;

use common::{prod, reference, rx, sens_term, term, tok};
use tabnas_bnf::{emit_grammar_spec, ConvertOptions, Element, Grammar, GrammarSpec};

// doc = x ; x = <a> t u / <b> u t ; t = "." "." ; u = "," ","
fn grammar(a: Element, b: Element) -> Grammar {
    Grammar::new(vec![
        prod("doc", vec![vec![reference("x")]]),
        prod(
            "x",
            vec![
                vec![a, reference("t"), reference("u")],
                vec![b, reference("u"), reference("t")],
            ],
        ),
        prod("t", vec![vec![sens_term("."), sens_term(".")]]),
        prod("u", vec![vec![sens_term(","), sens_term(",")]]),
    ])
}

fn depths(spec: &GrammarSpec) -> Vec<usize> {
    spec.rule["x"]
        .as_ref()
        .expect("x")
        .open
        .iter()
        .map(|o| o.s().unwrap_or("").split_whitespace().count())
        .collect()
}

fn parses(spec: &GrammarSpec, src: &str, relex: bool) -> bool {
    let mut options = tabnas::Options::default();
    options.lex.relex = relex;
    let mut parser = tabnas::Tabnas::with_options(options);
    spec.install(&mut parser).expect("install");
    parser.parse(src).is_ok()
}

fn opts(word_keywords: bool) -> ConvertOptions {
    ConvertOptions::tag("ct")
        .start("doc")
        .word_keywords(word_keywords)
}

#[test]
fn a_word_keyword_contests_a_longer_literal_that_continues_with_punctuation() {
    let spec = emit_grammar_spec(&grammar(sens_term("a"), sens_term("a-b")), &opts(true)).unwrap();
    assert_eq!(depths(&spec), [2, 2]);
    assert!(parses(&spec, "a-b,,..", true));
    assert!(parses(&spec, "a..,,", true));
    let word = emit_grammar_spec(&grammar(sens_term("a"), sens_term("ab")), &opts(true)).unwrap();
    assert_eq!(depths(&word), [1, 1], "a word continuation keeps the guard");
}

#[test]
fn case_insensitive_literals_contest_by_what_their_matchers_fold() {
    // A literal with no ASCII letter compiles to an exact fixed token, so
    // Σ and ς never meet; with ASCII letters the matcher folds case.
    let spec = emit_grammar_spec(&grammar(term("Σ"), term("ς")), &opts(false)).unwrap();
    assert_eq!(depths(&spec), [1, 1]);
    assert!(parses(&spec, "ς,,..", false));
    assert!(parses(&spec, "Σ..,,", false));
    let ascii = emit_grammar_spec(&grammar(term("a"), term("A-b")), &opts(false)).unwrap();
    assert_eq!(depths(&ascii), [2, 2]);
    assert!(parses(&ascii, "A-B,,..", true));
    assert!(parses(&ascii, "A..,,", true));
}

#[test]
fn a_regex_head_that_is_not_one_atom_contests_by_what_it_can_match() {
    let spec = emit_grammar_spec(&grammar(rx("a|b", ""), sens_term("b")), &opts(false)).unwrap();
    assert_eq!(depths(&spec), [2, 2]);
    assert!(parses(&spec, "b,,..", true));
    assert!(parses(&spec, "a..,,", true));
}

#[test]
fn the_any_token_contests_every_head() {
    let spec = emit_grammar_spec(&grammar(tok("#AA"), sens_term("b")), &opts(false)).unwrap();
    assert_eq!(depths(&spec), [2, 2]);
    assert!(parses(&spec, "b,,..", false));
    assert!(parses(&spec, "z..,,", false));
}

#[test]
fn a_regex_head_whose_first_character_the_coverage_cannot_name_contests() {
    // `\n`, `\t` and `\v` are one atom each, but `pattern_char_ranges`
    // declines to name what a control escape covers, so nothing says the
    // literal character is not what the pattern matches.
    for (pattern, literal) in [(r"\n", "\n"), (r"\t", "\t"), (r"\v", "\u{b}")] {
        let spec =
            emit_grammar_spec(&grammar(rx(pattern, ""), sens_term(literal)), &opts(false)).unwrap();
        assert_eq!(depths(&spec), [2, 2], "{pattern}");
    }
    let spec =
        emit_grammar_spec(&grammar(rx(r"\v", ""), sens_term("\u{b}")), &opts(false)).unwrap();
    assert!(parses(&spec, "\u{b}..,,", true));
    assert!(parses(&spec, "\u{b},,..", true));
}

#[test]
fn a_case_insensitive_head_beyond_ascii_contests_what_unicode_folding_lets_it_meet() {
    // `(?i)[Σ]` takes `ς`, and the coverage folds ASCII letters alone.
    let spec = emit_grammar_spec(&grammar(rx("[Σ]", "i"), sens_term("ς")), &opts(false)).unwrap();
    assert_eq!(depths(&spec), [2, 2]);
    for src in ["ς..,,", "Σ..,,", "ς,,.."] {
        assert!(parses(&spec, src, true), "{src}");
    }
    // Within ASCII the folded coverage stays exact: `(?i)[b]` meets no `a`.
    let ascii = emit_grammar_spec(&grammar(rx("[b]", "i"), sens_term("a")), &opts(false)).unwrap();
    assert_eq!(depths(&ascii), [1, 1]);
}

// `grammar` with tails no engine matcher can take (a number would swallow
// the `.` of `1..`): t = ";" ";" ; u = "!" "!".
fn semi(a: Element, b: Element) -> Grammar {
    Grammar::new(vec![
        prod("doc", vec![vec![reference("x")]]),
        prod(
            "x",
            vec![
                vec![a, reference("t"), reference("u")],
                vec![b, reference("u"), reference("t")],
            ],
        ),
        prod("t", vec![vec![sens_term(";"), sens_term(";")]]),
        prod("u", vec![vec![sens_term("!"), sens_term("!")]]),
    ])
}

#[test]
fn an_engine_token_contests_what_its_matcher_can_take_when_the_parser_asks() {
    // Negotiated lexing runs only the matchers that can produce the token
    // an alternative wants: the number matcher takes a leading digit, the
    // string matcher a quote, and the text matcher any text no fixed
    // literal claims. The four-token dispatch this replaced kept each of
    // these pairs apart; a one-token dispatch did not.
    let cases = [
        (tok("#NR"), sens_term("1"), "1"),
        (tok("#ST"), sens_term("'a'"), "'a'"),
        (tok("#TX"), term("let"), "let"),
        (tok("#NR"), tok("#TX"), "1"),
        (tok("#NR"), rx("[0-9]", ""), "1"),
        (tok("#TX"), rx("[a-z]+", ""), "let"),
    ];
    for (a, b, text) in cases {
        let spec = emit_grammar_spec(&semi(a, b), &opts(false)).unwrap();
        assert_eq!(depths(&spec), [2, 2], "{text}");
        assert!(parses(&spec, &format!("{text};;!!"), true), "{text};;!!");
        assert!(parses(&spec, &format!("{text}!!;;"), true), "{text}!!;;");
    }
    // The text matcher defers to a fixed literal, and an emitted grammar
    // lexes no values: those stay one token deep.
    for (a, b) in [
        (tok("#TX"), sens_term("let")),
        (tok("#VL"), sens_term("true")),
    ] {
        let spec = emit_grammar_spec(&semi(a, b), &opts(false)).unwrap();
        assert_eq!(depths(&spec), [1, 1]);
    }
}

#[test]
fn a_brace_escape_is_a_code_point_only_under_the_u_flag() {
    // Without `u` or `v`, `\u{1}` is `u` once in the JavaScript matcher
    // the canonical runtime emits for, not U+0001, so the emitted dispatch
    // treats the head as inexact. The depths are the emitted-spec
    // contract; the Rust regex crate reads `\u{1}` as a code point under
    // any flags, so the parse side is pinned in TypeScript alone.
    let legacy = emit_grammar_spec(&semi(rx(r"\u{1}", ""), sens_term("u")), &opts(false)).unwrap();
    assert_eq!(depths(&legacy), [2, 2]);
    let unicode =
        emit_grammar_spec(&semi(rx(r"\u{1}", "u"), sens_term("u")), &opts(false)).unwrap();
    assert_eq!(depths(&unicode), [1, 1]);
}

#[test]
fn an_escape_whose_code_point_cannot_be_read_is_not_an_exact_head() {
    // A head the coverage cannot read contests every head. `\p{L}` is a
    // property, not the letter `p`, so a class holding it meets `é`. The
    // escapes the regex crate refuses outright (`\u1`, `\x1`, `\cA`, a
    // digit escape) never reach the dispatcher here; the two readers pin
    // them (`an_escape_is_one_code_point_only_when_it_can_be_read` in
    // `src/ranges.rs`), and TypeScript pins the dispatch.
    let property =
        emit_grammar_spec(&semi(rx(r"[\p{L}]", "u"), sens_term("é")), &opts(false)).unwrap();
    assert_eq!(depths(&property), [2, 2]);
    // Four hex digits are the code point they spell.
    let exact = emit_grammar_spec(&semi(rx(r"\u0041", ""), sens_term("u")), &opts(false)).unwrap();
    assert_eq!(depths(&exact), [1, 1]);
}

#[test]
fn an_escape_is_read_as_the_regex_crate_reads_it() {
    // This port compiles every matcher with the `regex` crate, so an
    // escape is read as that crate reads it, not as JavaScript does. `\a`
    // is BEL, which a JavaScript matcher reads as the letter `a`, and
    // `\U` spells a code point in eight hex digits or in braces. Read as
    // the letter after the backslash, each of these heads was held apart
    // from a literal BEL it takes, and the literal's branch was never
    // reached. Mirrors TestContestEscapeIsReadAsRE2ReadsIt in
    // go/contest_test.go.
    for pattern in [
        r"\a",
        r"[\a]",
        r"\a+",
        r"[\U00000007]",
        r"[\U{7}]",
        r"\x{7}",
    ] {
        let spec =
            emit_grammar_spec(&semi(rx(pattern, ""), sens_term("\x07")), &opts(false)).unwrap();
        assert_eq!(depths(&spec), [2, 2], "{pattern}");
        assert!(parses(&spec, "\x07;;!!", true), "{pattern}");
        assert!(parses(&spec, "\x07!!;;", true), "{pattern}");
    }
}
